Docs/push notifications (#2611)
esl/MongooseIM: doc/user-guide/push-notifications/Push-notifications.md

notifications for FCM (**F**irebase **C**loud **M**essaging) and APNS (**A**pple **P**ush
**N**otification **S**ervice) providers. And it's easy to extend it to any other protocols of your
choice.

You might also want to read about the push notification's [client side configuration][client-side].

All push notification mechanisms use [mod_event_pusher_push][] as a backend implementation,
read [the relevant documentation][mod_event_pusher] to know more about it.

### [XEP-0357 Push Notifications][XEP-0357]

Server side push notification support is fully compliant with [XEP-0357 Push Notifications][XEP-0357],
which defines several components that need to work together in order to provide clients with working
push notifications. However, there's just one non-optimal detail required by the aforementioned XEP:
that push notifications being a PubSub service — [we can do better than that][pubsub-less].

If you're already familiar with the workings of [XEP-0357][], make sure to have a look at
[our PubSub-less enhancement][pubsub-less].

As it is always said, one picture is worth a thousand words:

 ![](PushNotifications-XEP-overview.png)

> Who does what is highly configurable. You may use MongooseIM as the XMPP server clients connect to,

Reviewer: ```suggestion
You may use MongooseIM as the XMPP server clients connect to
```

and send the push XMPP stanzas to a different server that will take care of the push business;
or you might use MongooseIM as the remote XMPP-PubSub server that does such business. Note that the
XEP doesn't enforce the push IQ stanza format, so whichever setup is used, you need to take care of
the producing and processing of these stanzas.

You might also use MongooseIM as both, or you might even do both things within a single MongooseIM
node (the most common setup!). Or, for the best performance, you might just skip that PubSub node
altogether. While the whole setup can be incredibly extensible, we see the following straightforward
uses of it.

### XEP-0357 compliant with local PubSub

 ![](PushNotifications-MIM_with_local_pubsub.png)

This is, historically, the most common setup.
It allows your clients to enable push notifications via a local PubSub,
and the IQ stanza is routed internally.

A direct connection to a push service (e.g. MongoosePush) must be configured on the same MongooseIM

> node, check out [this][WithMongoosePush] tutorial on how to setup [MongoosePush][].

Reviewer: ```suggestion
node. Check out [this][WithMongoosePush] tutorial on how to setup [MongoosePush][].
```

They only have to provide the virtual PubSub address in the enable stanza, and node name can be
anything unique. In order to ensure uniqueness the APNS/FCM token can be used. Note that the token
must be provided [as a publish option](https://xmpp.org/extensions/xep-0357.html#example-9) anyway.

A direct connection to a push service (e.g. MongoosePush) must be configured on the same MongooseIM
node, check out [this][WithMongoosePush] tutorial on how to setup [MongoosePush][].

```Erlang
{mod_event_pusher, [
    {backends, [
        {push, [
            {backend, mnesia}, % optional
            {wpool, [{workers, 200}]}, % optional
            {plugin_module, mod_event_pusher_push_plugin_defaults}, % optional
            {virtual_pubsub_hosts, ["virtual.@HOST@"]}
        ]}
    ]}
]}
```

#### Advantages
* No need to use PubSub at all
* More efficient (PubSub has a considerable impact on heavily loaded systems)
* Simpler client-side usage — Read about the [client side configuration here][client-side].

#### Drawbacks
* If the client application is build to create the push PubSub node, this might require a migration

>   for such client — as he attempts to create the node, the server will answer with an IQ error

Reviewer: ```suggestion
  for such client — as they attempt to create the node, the server will answer with an IQ error
```

>   stanza if the domain is not shadowed (using `virtual_pubsub_hosts`). If migrating the client side

Reviewer 2: this is completely incorrect and must be changed. is pubsub is not configured server responds on node creation request with IQ error even if domain is shadowed! shadowing is not solving the this problem hence the need for migration tricks!

All push notification mechanisms use [mod_event_pusher_push][] as a backend implementation,
read [this][mod_event_pusher] to know more about it.

While the whole setup can be incredibly extensible, we see three straightforward uses of it.

### MongooseIM as a PubSub-less XMPP server

 ![](PushNotifications-no_pubsub.png)

PubSub is completely bypassed and the clients don't need to create a push node.
They only have to provide the virtual PubSub address.

A direct connection to a push service (e.g. MongoosePush) must be configured on the same MongooseIM
node, checkout [this][WithMongoosePush] tutorial on how to setup [MongoosePush][].

If compatibility with client libraries is not an issue, this is a really good approach.
PubSub is completely bypassed and the clients don't need to create a push node — if they attempt to
do so, the server would respond with an error stanza, hence the incompatibility with [XEP-0357][].
They only have to provide the virtual PubSub address.

```Erlang
{mod_event_pusher, [
    {backends, [
        {push, [
            {backend, mnesia}, % optional
            {wpool, [{workers, 200}]}, % optional

>             {plugin_module, mod_event_pusher_push_plugin_defaults}, % optional, read more [here][mod_event_pusher_push_plugin]

Reviewer 3: Are you sure the link will work in the code block?

Author: Damn, good catch, it won't! 😅

support for them. When used together with [MongoosePush][], you get out-of-the-box push
notifications for FCM (**F**irebase **C**loud **M**essaging) and APNS (**A**pple **P**ush
**N**otification **S**ervice) providers. And it's easy to extend it to any other protocols of your
choice.

Server side push notification support is fully compliant with [XEP-0357 Push Notifications][XEP-0357],
which defines several components that need to work together in order to provide clients with working
push notifications. However, there's just one non-optimal detail required by the aforementioned XEP:
that push notifications being a PubSub service — [we can do better than that][pubsub-less].

If you're already familiar with the workings of [XEP-0357][], make sure to have a look at
[our pubsub-less enhancement][pubsub-less].

You might also want to read about the push notification's [client side configuration][client-side].

All push notification mechanisms use [mod_event_pusher_push][] as a backend implementation,
read [this][mod_event_pusher] to know more about it.

While the whole setup can be incredibly extensible, we see the following straightforward uses of it.

> ### XEP-0357 compliant with local PubSub

Reviewer 2: please arrange the sections in the next order:

1. XEP-0357 compliant with remote PubSub XMPP server
   - note here that the format of the push IQ is can be adjusted via the mod_event_pusher_push custom-made plugin module.
   - also mention that all the provided plugins are generating IQ that that is specific to our own `push PubSub node` implementation
2. XEP-0357 compliant with MIM used as remote PubSub XMPP server
   - state here one more time that we expect push notification IQ to have specific (not declared in XEP) custom fields, e.g. device token. So if someone wants to use with 3rd party XMPP server, he must ensure that this 3rd party server generates push IQ in that specific format. 
3. XEP-0357 compliant when PubSub is configured at the same MIM cluster and IQ is routed internally.
   - MB mention here that this was the most common use-case. but it's senseless and ineffective to generate extra routing, and in the most cases PubSub was the bottleneck of the whole setup. so we came up to an idea of PubSub-less setup.
4. PubSub-less configuration.

Reviewer 2: actually the first and the second point can be replaced with some general XEP overview with an abstract diagram.

Author: Done in 818f62403

            {wpool, [{workers, 200}]}, % optional
            {plugin_module, mod_event_pusher_push_plugin_defaults} % optional
        ]}
    ]}
]}
```

### XEP-0357 compliant with remote PubSub

 ![](PushNotifications-MIM_with_remote_pubsub.png)

In this case, clients must provide a PubSub service domain that exists on another XMPP cluster.
Notifications will be delivered over XMPP Federation to the remote server(s), so it all comes down to a
[federation configuration][federation].

#### Advantages
* The whole responsibility is outsourced, MongooseIM will solely route the stanza abroad.
* No need to worry about pubsub-less/pubsub-full or [XEP-0357][] compliance, that's precisely
  outsourced.

#### Drawbacks
* It still all depends on how the federated server is configured. It can be a transparent custom
  service, or another MongooseIM node with any of the previous configurations.

If MongooseIM is used as the remote PubSub node, the architecture then looks like this:
 ![](PushNotifications-MIM_as_remote_pubsub.png)

The components that make push notifications possible in MongooseIM in the PubSub-full setup
add up to the following architecture:

> ![](component-architecture-with-PubSub.png)

Reviewer 3: There is a typo in the file the link refers to. It contains `achitecture` instead of `architecture`, so the link does not work.

do so, the server would respond with an error stanza, hence the incompatibility with [XEP-0357][].
They only have to provide the virtual PubSub address.

```Erlang
{mod_event_pusher, [
    {backends, [
        {push, [
            {backend, mnesia}, % optional
            {wpool, [{workers, 200}]}, % optional
            {plugin_module, mod_event_pusher_push_plugin_defaults}, % optional
            {virtual_pubsub_hosts, ["virtual.@HOST@"]}
        ]}
    ]}
]}
```

#### Advantages
* No need to use PubSub at all
* More efficient (PubSub has a considerable impact on heavily loaded systems)
* Simpler client-side usage

#### Drawbacks
* It is only partially XEP-0357 compliant, so 3rd party clients most probably won't work even in
  private deployments — they need to be adapted
* Not suitable for deployments that require connections from 3rd party client applications.
  Notifications won't work at all in such case

### XEP-0357 compliant + virtual hosts

This is the most versatile setup. Allows clients to use all push notifications configurations:

The components that make push notifications possible in MongooseIM add up to the following architecture:

> ![](component-architecture-without-PubSub.png)

Reviewer 3: This file has a typo with `achitecture` as well.

* PubSub-less
* Push service handled locally via XEP-0357
* Push service handled by another cluster via XEP-0357

```Erlang
{mod_pubsub, [{plugins, [<<"push">>]}]}, % mandatory minimal config

{mod_event_pusher, [
    {backends, [
        {push, [
            {backend, mnesia}, % optional
            {wpool, [{workers, 200}]}, % optional
            {plugin_module, mod_event_pusher_push_plugin_defaults}, % optional
            {virtual_pubsub_hosts, ["virtual.@HOST@"]}
        ]}
    ]}
]}
```

#### Virtual PubSub hosts

These domains will shadow any identical domain configured for pubsub, stealing any notification
published to it. It enables easy migration from PubSub-full deployments to PubSub-less variants.
Read more in [mod_event_pusher_push_virtual_hosts][].

#### Advantages

> * Versatility: pubsub-less and pubsub-full mechanisms can be configured with different domains and

Reviewer 3: Minor detail, and I believe most files in the documentation lack consistency with this word, but in this line we have `pubsub-less` and in the line 142 there is `PubSub-less`. It just looks weird spelled differently in the same paragraph.

Author: Fun fact, the XEP is also mixing both spellings in the same paragraphs: https://xmpp.org/extensions/xep-0357.html 😂 
But I'll try going to the camel-case version anyway xD

## How to set up Push Notifications

Push notifications are the bread and butter of the modern mobile experience, and MongooseIM has
support for them. When used together with [MongoosePush][], you get out-of-the-box push
notifications for FCM (**F**irebase **C**loud **M**essaging) and APNS (**A**pple **P**ush
**N**otification **S**ervice) providers. And it's easy to extend it to any other protocols of your
choice.

> You might also want to read about the push notification's [client side configuration][client-side].

Reviewer 3: I believe it's my last complaint, but this file is only available through this one link, and when you enter it, you are nowhere in the docs tree (it is not listed in the files structure on the left). I don't have any idea how to fix this problem.
Perhaps a link in line 105 would at least make the chance of finding this file a little higher.

```erlang
{mod_pubsub, [{plugins, [<<"push">>]}]}, % mandatory minimal config
{mod_event_pusher, [
    {backends, [
        {push, [
            {backend, mnesia}, % optional
            {wpool, [{workers, 200}]}, % optional
            {plugin_module, mod_event_pusher_push_plugin_defaults} % optional
        ]}
    ]}
]}
```

#### Advantages
* Completely XEP-0357 compliant
* Most likely compatible with most 3rd party client libraries

#### Drawbacks
* Less efficient (PubSub has a considerable impact on heavily loaded systems)
* More load within a single node
* Harder to customise

### MongooseIM as a PubSub-less XMPP server

 ![](PushNotifications-no_pubsub.png)

If compatibility with client libraries is not an issue, this is a really good approach.
PubSub is completely bypassed and clients don't need to create a push node — if they attempt to
do so, the server would respond with an error stanza, hence the incompatibility with [XEP-0357][].
They only have to provide the virtual PubSub address in the enable stanza, and node name can be
anything unique. In order to ensure uniqueness the APNS/FCM token can be used. Note that the token

> must be provided [as a publish options](https://xmpp.org/extensions/xep-0357.html#example-9) anyway.

Reviewer 3: `as a publish option` or `as one of the publish options`? Or is it intentional?

A direct connection to a push service (e.g. MongoosePush) must be configured on the same MongooseIM

> node, checkout [this][WithMongoosePush] tutorial on how to setup [MongoosePush][].

Reviewer 3: `check out` instead of `checkout`

Reviewer 3: Actually this sentence appears three times in this file.

```Erlang
{mod_event_pusher, [
    {backends, [
        {push, [
            {backend, mnesia}, % optional
            {wpool, [{workers, 200}]}, % optional
            {plugin_module, mod_event_pusher_push_plugin_defaults}, % optional
            {virtual_pubsub_hosts, ["virtual.@HOST@"]}
        ]}
    ]}
]}
```

#### Advantages
* No need to use PubSub at all
* More efficient (PubSub has a considerable impact on heavily loaded systems)
* Simpler client-side usage

#### Drawbacks

> * It is only partially XEP-0357 compliant, so 3rd party clients most probably won't work even in

Reviewer 2: that's not true, XEP doesn't declare creation of the node as mandatory step. it declares only that PushService is represented by the `push PubSub node`. see [Business Rules](https://xmpp.org/extensions/xep-0357.html#sect-idm46544864749808) and [Enabling Notifications](https://xmpp.org/extensions/xep-0357.html#enabling) paragraphs

```
The App Server allocates or reuses a node on the App Server's XMPP Push Service
```

Reviewer 2: also according to the ASCII diagram in XEP, it doesn't have to be done via XMPP client.

  private deployments — they need to be adapted
* Not suitable for deployments that require connections from 3rd party client applications.

>   Notifications won't work at all in such case

Reviewer 3: I believe it should be `in such a case`?

Author: Doesn't really matter I think
https://english.stackexchange.com/questions/198941/in-such-case-and-in-such-a-case

### XEP-0357 compliant + virtual hosts

> This is the most versatile setup, a compromise of both worlds.

Reviewer 3: `compromise between` sounds better in my opinion
